### PR TITLE
Draft: Spam signal

### PR DESCRIPTION
**Pull request by rsdeus**

### CODE REVIEW

File: app/controllers/decidim/spam_signal/spam-signal_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require_dependency "decidim/spam_signal/application_controller"
```
Comment (froger) [inline app/controllers/decidim/spam_signal/spam-signal_controller.rb:1]: I know it's boilerplate, but you can remove this file ;)

File: app/jobs/decidim/spam_signal/spam_signal_job.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      
+      def perform(logger = Rails.logger)
+        # @logger = logger
+        invoke!
+      end
+
+      private
+
+        def invoke!
+          puts "Init Job"
```
Comment (froger) [inline app/jobs/decidim/spam_signal/spam_signal_job.rb:18]: Avoid using puts. If not in a proper logger, this becomes hard to track.

File: decidim-spam_signal.gemspec [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  s.files = Dir["{app,config,lib}/**/*", "LICENSE.md", "Rakefile", "README.md"]
+
+  s.require_paths = ["lib"]
+  s.add_dependency "decidim-core", Decidim::SpamSignal.version
+  s.add_dependency "decidim", Decidim::SpamSignal.version
+  s.add_dependency "rails"
+  s.add_development_dependency "rspec"
+  s.add_development_dependency "rspec_junit_formatter"
+  s.add_development_dependency "decidim-dev", Decidim::SpamSignal.version
+  s.add_development_dependency "decidim-consultations", Decidim::SpamSignal.version
```
Comment (froger) [inline decidim-spam_signal.gemspec:28]: I think you need only decidim-core as dep in your case, you won't use other module than report and users.

File: lib/decidim/spam_signal/engine.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+require "rails"
+require "decidim/core"
+
+module Decidim
+  module SpamSignal
+    # This is the engine that runs on the public interface of spam_signal.
+    class Engine < ::Rails::Engine
+      isolate_namespace Decidim::SpamSignal
+
+      initializer "decidim_spam_signal.assets" do |app|
```
Comment (froger) [inline lib/decidim/spam_signal/engine.rb:12]: Boilerplate, you can remove the initializer.

File: app/jobs/decidim/spam_signal/spam_signal_job.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+          spam_signal = Decidim::SpamSignal::AppSpamSignal.new
```
Comment (froger) [inline app/jobs/decidim/spam_signal/spam_signal_job.rb:19]: You can move the whole code of `Decidim::SpamSignal::AppSpamSignal` in the job. 
`.perform_later` goes in sidekiq, `perform_now` execute the code directly. No need for separation, we are on the same concern here.

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        end
+      end
+
+      private 
+        def take_or_create_user_bot
+            
+          raise "Organization doesn't exists" unless @organization = Decidim::Organization.first
+
+          user_bot = Decidim::User.find_by(email: @user_bot_email) || Decidim::User.create!(
+            email: @user_bot_email,
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:26]: avoid using `@` variable for other thing than initializers. use `attr_reader` and `attr_writter` instead.

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+          unless user_signaled?(user) || user.blocked?
+            signal_spammer(user) if user_about_uri?(user) && user_about_text?(user)
+          end
+        end
+      end
+
+      private 
+        def take_or_create_user_bot
+            
+          raise "Organization doesn't exists" unless @organization = Decidim::Organization.first
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:23]: I see an immediate optimisation here: 

```
   raise "Organization doesn't exists" unless @organization = Decidim::Organization.first
```

A cleaner way would be to make an accessor: 
```
def organization
    Decidim::Organization.first! # will raise NotFound if not exists — note the "!"
end
```

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+          user_bot = Decidim::User.find_by(email: @user_bot_email) || Decidim::User.create!(
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:25]:
```suggestion
          user_bot = Decidim::User.find_or_create_by_email(email: @user_bot_email) do |usr|
             usr.name = "bot"
             usr.nickname = "bot"
             usr.password = usr.password_confirmation = Devise.friendly_token
             usr.organization = organization
             usr.confirmed_at = Time.current
             usr.locale =I18n.default_locale
             usr.admin = false
          end
```

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            email: @user_bot_email,
+            name: 'bot',
+            nickname: 'bot',
+            password: 'changeme01',
+            password_confirmation: 'changeme01',
+            organization: organization,
+            confirmed_at: Time.current,
+            locale: I18n.default_locale,
+            admin: true,
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:34]: Unnecessary to have a bot user with admin rights ;)

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            tos_agreement: true,
+            accepted_tos_version: organization.tos_version,
+            admin_terms_accepted_at: Time.current
+          )
+
+          user_bot.update(:blocked => false) if user_bot.blocked?
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:40]:
```suggestion
          user_bot.update(blocked: false) if user_bot.blocked?
```

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+          user_bot
+          
+        end
+
+        def take_regex
+          regex = Regexp.union(regex_string)
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:46]: no need to assign a local var. 
```suggestion
          Regexp.union(regex_string)
```

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  module SpamSignal
+    class AppSpamSignal
+
+      def run!
+        @user_bot_email = ENV.fetch("USER_BOT_EMAIL", nil)
+        @uncheckable_email_domain = ENV.fetch("UNCHECKABLE_EMAIL_DOMAIN", nil)
+        @user_bot = take_or_create_user_bot
+        @regex = /#{take_regex}/i
+        
+        users_to_check.each do |user|
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:13]: How about using [`Decidim::UserModeration`](https://github.com/decidim/decidim/blob/4f8f8c47e340c95281712134652b298d28e31804/decidim-core/app/models/decidim/user_moderation.rb). 

Something like

```ruby
Decidim.UserModeration.unblocked.where(decidim_users: {email: <stuff>}) .includes(:reports).where("decidim_reports.moderation_id": nil).each do |moderation|
 user = moderation.user
 # do your checks
end
```

It won't really optimise the query (still in O(n)), but it's a bit better I think

File: lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+          @report = Decidim::UserReport.create!(
+            moderation: @moderation, 
+            user: @user_bot,
+            reason: 'spam',
+            details: 'Signeled by bot'
+          )
+        end
+
+        def update_report_count!
+          @moderation.update!(report_count: @moderation.report_count + 1)
```
Comment (froger) [inline lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb:104]: that's very strage, are you sure you need that?

**Review by froger — CHANGES_REQUESTED** [review]
Hi @rsdeus that's cool first draft of the feature thanks <3
I have some minors changes, I saw nothing that would break things. 

To merge, I need you to 

* write unit tests
* remove unused dependancies in the Gemfile
* remove the `@` instance attribute everywhere and use attr_reader.

The rest, that's just feedback, integrates as you feel.

File: app/jobs/decidim/spam_signal/spam_signal_job.rb (hunk 5 above)
Comment (rsdeus) [reply app/jobs/decidim/spam_signal/spam_signal_job.rb:19]: Sorry, I didn't understand, I can get off the code from lib and put it directly on the job?

Comment (froger) [reply app/jobs/decidim/spam_signal/spam_signal_job.rb:19]: Yes, that what I meant but with the changes we discussed off-line I think most of this PR will be deprecated.

**Pull request closed without merging**